File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.java
import java.util.*;

class Solution {
    public int[][] asSoonAsPossibleMinimum;
    public int N;
        
    public int getMinimumPathSum(int[][] matrix, int y, int x) {
        if (y == N - 1) {
            return matrix[y][x];
        }
        
        if (asSoonAsPossibleMinimum[y][x] != Integer.MAX_VALUE) {
            return asSoonAsPossibleMinimum[y][x];
        }
        
        asSoonAsPossibleMinimum[y][x] = 0;
        int bestChoice = Integer.MAX_VALUE;
        
        for (int i = -1; i <= 1; i++) {
            int newY = y + 1;
            int newX = x + i;
            
            if (!outOfRange(newY, newX)) {
                bestChoice = Math.min(bestChoice, getMinimumPathSum(matrix, newY, newX));
            }
        }
        
        asSoonAsPossibleMinimum[y][x] = matrix[y][x] + bestChoice;
        return asSoonAsPossibleMinimum[y][x];
    }
    
    public boolean outOfRange(int y, int x) {
        return (y < 0 || y >= N) || (x < 0 || x >= N);
    }
    
    public int minFallingPathSum(int[][] matrix) {
        N = matrix.length;
        asSoonAsPossibleMinimum = new int[N][N];
        
        for (int i = 0; i < N; i++) {
            Arrays.fill(asSoonAsPossibleMinimum[i], Integer.MAX_VALUE);
        }
        
        int answer = Integer.MAX_VALUE;
        for (int i = 0; i < N; i++) {
            answer = Math.min(answer, getMinimumPathSum(matrix, 0, i));
        }
        
        return answer;
    }
}

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.ts
function isOutOfRanage(x: number, range) {
    return x < 0 || x >= range;
}

function minFallingPathSum(matrix: number[][]): number {
    const dx = [-1, 0, 1];
    let dp = [];
    
    for (let i = 0; i < matrix.length; i++) {
        dp.push([]);
        dp[0].push(matrix[0][i]);
    }

    for (let i = 1; i < matrix.length; i++) {
        dp.push([]);

        for (let j = 0; j < matrix[i].length; j++) {
            let minimumValue = Number.MAX_VALUE;
            for (let dir = 0; dir < 3; dir++) {
                const nx = j + dx[dir];

                if (isOutOfRanage(nx, matrix[i].length)) {
                    continue;
                }

                minimumValue = Math.min(minimumValue, dp[i - 1][nx]);
            }
            dp[i].push(minimumValue + matrix[i][j]);
        }
    }
    
    let min = Number.MAX_VALUE;
    for (let i = 0; i < matrix[0].length; i++) {
        min = Math.min(min, dp[matrix.length - 1][i]);
    }
    return min;
}
